From ab503531adab7f5287f99ce13d7c79c7cd93375c Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Thu, 29 Aug 2019 11:18:45 +0800
Subject: [PATCH 19/81] miext: misyncfd: Fix memory leak

The private resource is never freed, which would cause memory leak.

Add CloseScreen() wrapper and free it there;

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 miext/sync/misyncfd.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/miext/sync/misyncfd.c b/miext/sync/misyncfd.c
index 92f3b2294..f211fbd75 100644
--- a/miext/sync/misyncfd.c
+++ b/miext/sync/misyncfd.c
@@ -34,6 +34,7 @@ static DevPrivateKeyRec syncFdScreenPrivateKey;
 
 typedef struct _SyncFdScreenPrivate {
     SyncFdScreenFuncsRec        funcs;
+    CloseScreenProcPtr          CloseScreen;
 } SyncFdScreenPrivateRec, *SyncFdScreenPrivatePtr;
 
 static inline SyncFdScreenPrivatePtr sync_fd_screen_priv(ScreenPtr pScreen)
@@ -66,6 +67,17 @@ miSyncFDFromFence(DrawablePtr pDraw, SyncFence *pFence)
     return (*priv->funcs.GetFenceFd)(pDraw->pScreen, pFence);
 }
 
+static Bool
+miSyncFdCloseScreen(ScreenPtr pScreen)
+{
+    SyncFdScreenPrivatePtr      priv = sync_fd_screen_priv(pScreen);
+
+    pScreen->CloseScreen = priv->CloseScreen;
+    free(priv);
+
+    return (*pScreen->CloseScreen) (pScreen);
+}
+
 _X_EXPORT Bool miSyncFdScreenInit(ScreenPtr pScreen,
                                   const SyncFdScreenFuncsRec *funcs)
 {
@@ -93,6 +105,9 @@ _X_EXPORT Bool miSyncFdScreenInit(ScreenPtr pScreen,
 
     priv->funcs = *funcs;
 
+    priv->CloseScreen = pScreen->CloseScreen;
+    pScreen->CloseScreen = miSyncFdCloseScreen;
+
     dixSetPrivate(&pScreen->devPrivates, &syncFdScreenPrivateKey, priv);
 
     return TRUE;
-- 
2.38.5

